cpp/test/prims/merge_labels.cu

```diff
@@ -0,0 +1,157 @@
+/*
+ * Copyright (c) 2020-2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <gtest/gtest.h>
+#include <label/merge_labels.cuh>
+
+#include <raft/cudart_utils.h>
+#include <cuml/common/cuml_allocator.hpp>
+#include "test_utils.h"
+
+// #include <iostream>
+// #include <limits>
+#include <vector>
+
+namespace MLCommon {
+namespace Label {
+
+template <typename Index_>
+struct MergeLabelsInputs {
+  Index_ N;
+  std::vector<Index_> labelsA;
+  std::vector<Index_> labelsB;
+  std::vector<uint8_t> mask;  // to avoid std::vector<bool> optimization
+  std::vector<Index_> expected;
+};
+
+template <typename Index_>
+class MergeLabelsTest
+  : public ::testing::TestWithParam<MergeLabelsInputs<Index_>> {
+ protected:
+  void SetUp() override {
+    params = ::testing::TestWithParam<MergeLabelsInputs<Index_>>::GetParam();
+
+    CUDA_CHECK(cudaStreamCreate(&stream));
+    std::shared_ptr<deviceAllocator> alloc(
+      new raft::mr::device::default_allocator);
+
+    raft::allocate(labelsA, params.N);
+    raft::allocate(labelsB, params.N);
+    raft::allocate(expected, params.N);
+    raft::allocate(R, params.N);
+    raft::allocate(mask, params.N);
+    raft::allocate(m, 1);
+  }
+
+  void Run() {
+    params = ::testing::TestWithParam<MergeLabelsInputs<Index_>>::GetParam();
+
+    raft::update_device(labelsA, params.labelsA.data(), params.N, stream);
+    raft::update_device(labelsB, params.labelsB.data(), params.N, stream);
+    raft::update_device(expected, params.expected.data(), params.N, stream);
+    raft::update_device(mask, reinterpret_cast<bool *>(params.mask.data()),
+                        params.N, stream);
+
+    merge_labels(labelsA, labelsB, mask, R, m, params.N, stream);
+
+    cudaStreamSynchronize(stream);
+    ASSERT_TRUE(raft::devArrMatch<Index_>(expected, labelsA, params.N,
+                                          raft::Compare<Index_>()));
+  }
+
+  void TearDown() override {
+    CUDA_CHECK(cudaFree(labelsA));
+    CUDA_CHECK(cudaFree(labelsB));
+    CUDA_CHECK(cudaFree(expected));
+    CUDA_CHECK(cudaFree(R));
+    CUDA_CHECK(cudaFree(mask));
+    CUDA_CHECK(cudaFree(m));
```

Comment:
We might even want to consider just using the rmm device vector directly since we've discussed removing `device_buffer` altogether at some point soon

Reply:
I did not give much thought to that as it was a refactoring of existing tests. Do you think the change belongs to this PR?

Reply:
I have changed to `rmm::device_vector`, but I'm not really satisfied by this nor `rmm::device_uvector` for the following reasons:

- `rmm::device_vector` doesn't give a raw pointer directly, I had to use `thrust::raw_pointer_cast`.
- `rmm::device_uvector`, which is more appropriate because we don't need the initialization, has no default constructor, and it expects a stream in the constructor, which is only created in `SetUp` in our case.

Any idea to use them better than what I did in my latest commit?

Reply:
Instead of `SetUp`, [you can have a constructor](https://github.com/Macaulay2/googletest-1/blob/master/googletest/docs/FAQ.md#should-i-use-the-constructordestructor-of-the-test-fixture-or-the-set-uptear-down-function) `MergeLabelsTest`, that way you can construct a `device_uvector`.